add junit xml reporter

dougiechong opened the pull request:
creates a reporter that inherits from ut collecting reporter.
takes the failures/successes and parses by the arrow.
Report comes out in XML format found in the comment of the class

Closes #56 

<!--- Briefly describe your changes in the area below. Mention any issues this PR closes. -->

## Description
<!--- If necessary, go into depth of what this pull request is doing. -->

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- Feel free to remove whole sections, not points within the sections, that do not apply -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] **I am adding a new reporter.**
  - [x] I have included documentation with at least one example and description.
  - [x] I have added tests in `Tests/UnitTests/Reporters`.
  - [x] I have added a note to `CHANGELOG.md`.

## Contributing

- [x] I have read and agree to the [Contributor Agreement](https://github.com/sassoftware/jsl-hamcrest/blob/master/ContributorAgreement.txt)

<!--- Replace your name and e-mail below -->
Signed-off-by: Douglas Chong <Douglas.chong@predictum.com>

emccorkle committed:
Always show Log Tab for Bottom Test Runner.

emccorkle committed:
Don't show embedded log by default.

JustinChilton committed:
Add instance of matcher.

emccorkle committed:
Simplified matrix equality

from Vince Faller

emccorkle committed:
Allow anything for the assertion label.

Also, ut test should return code should match
information given to the reporter.

emccorkle committed:
Clarified StreamingLogReporter failure message.

JustinChilton committed:
Initial set of tests for JSL-Hamcrest.

JustinChilton committed:
Made AllOf mismatch wording less confusing.

JustinChilton committed:
Allow registering assertions for ut test.

JustinChilton committed:
Move matcher/LRE functions from Core to Utils.

JustinChilton committed:
Include JSL files recursively.

JustinChilton committed:
Mangle assertion locals to reduce name collisions.

JustinChilton committed:
Delete namespaces/classes in delete sym difference

emccorkle committed:
Bump to v1.1.0

emccorkle committed:
Switched to factory functions for Reporters.

emccorkle committed:
Refactored 'this obj' to 'self'.

emccorkle committed:
Allow unregistering matchers.

emccorkle committed:
Added basic tests for Addin reporters.

emccorkle committed:
Provide install folder for custom objects.

Both for Matchers and for Reporters.

emccorkle committed:
Do not overwrite ut global reporter.

jefurbee committed:
Initial commit

JustinChilton committed:
add v1.0.1

JustinChilton committed:
Merge pull request #9 from emccorkle/bump-version

Bump to v1.1.0

JustinChilton committed:
Merge pull request #10 from emccorkle/instance-of

Add instance of matcher.

JustinChilton committed:
Merge pull request #11 from emccorkle/add-tests

Initial set of tests for JSL-Hamcrest.

JustinChilton committed:
Merge pull request #12 from emccorkle/reporter-factories

Switched to factory functions for Reporters.

JustinChilton committed:
Merge pull request #13 from emccorkle/this-obj

Refactored 'this obj' to 'self'.

JustinChilton committed:
Merge pull request #14 from emccorkle/include-recursive

Include JSL files recursively.

JustinChilton committed:
Merge pull request #15 from emccorkle/test-case-label

Allow anything for the assertion label.

JustinChilton committed:
Merge pull request #16 from emccorkle/del-namespaces

Delete namespaces/classes in delete sym difference

JustinChilton committed:
Merge pull request #17 from emccorkle/assert-sym-collisions

Mangle assertion locals to reduce name collisions.

JustinChilton committed:
Merge pull request #28 from emccorkle/embed-log

Embed log

JustinChilton committed:
Merge pull request #29 from emccorkle/unregister-matchers

Allow unregistering matchers.

JustinChilton committed:
Merge pull request #30 from emccorkle/matrix-eq

Simplified matrix equality

JustinChilton committed:
Merge pull request #31 from emccorkle/allof-mismatch

Made AllOf mismatch wording less confusing.

JustinChilton committed:
Merge pull request #32 from emccorkle/streaming-reporter-expr-fail

Clarified StreamingLogReporter failure message.

JustinChilton committed:
Merge pull request #33 from emccorkle/addin-reporter-tests

Added basic tests for Addin reporters.

JustinChilton committed:
Merge pull request #34 from emccorkle/custom-folders

Provide install folder for custom objects.

JustinChilton committed:
Merge pull request #35 from emccorkle/move-utils

Move matcher/LRE functions from Core to Utils.

JustinChilton committed:
Merge pull request #36 from emccorkle/no-overwrite-reporter

Do not overwrite ut global reporter.

JustinChilton committed:
Merge pull request #37 from emccorkle/register-assertions

Allow registering assertions for ut test.

JustinChilton committed:
General MessageMatcher improvements. Add enabled matcher.

JustinChilton committed:
Update changelog with changes for this PR.

JustinChilton committed:
Add Globals Section comment to fix doc linking

emccorkle committed:
Merge pull request #45 from JustinChilton/instance-of-globals-section

Add Globals Section comment to fix doc linking for ut instance of

emccorkle committed:
Merge pull request #42 from JustinChilton/master

General MessageMatcher improvements. Add enabled matcher.

JustinChilton committed:
Add ut assert value

emccorkle committed:
Fixes broken ut mock reporter factory function.

emccorkle committed:
Merge pull request #47 from JustinChilton/assert-value

Add ut assert value

emccorkle committed:
Merge pull request #49 from emccorkle/mock-reporter-tests-broken

Fixes broken ut mock reporter factory function.

JustinChilton committed:
Add pull request template.

JustinChilton committed:
Add contributing agreement. Fix typo/wording.

emccorkle committed:
Merge pull request #50 from JustinChilton/pull-request-template

Add pull request template.

JustinChilton committed:
Add expr eval protection to Char/Size Matchers

JustinChilton committed:
Merge pull request #53 from JustinChilton/expr-eval-protection

Add expr eval protection to Char/Size Matchers

JustinChilton committed:
Add link to docs and fix title

JustinChilton committed:
Merge pull request #55 from JustinChilton/readme-docs-update

Add link to docs and fix title in README.md

dougiechong committed:
add junit xml reporter
creates a reporter that inherits from ut collecting reporter.
takes the failures/successes and parses by the arrow.
Report comes out in XML format found in the comment of the class

dougiechong commented:
I do not see the folder "Tests/UnitTests/Reporters"

emccorkle commented:
Sorry about that. The `Tests/UnitTests/Reporters` folder will have to be added with this PR. Just create it and add a test file. Any files contained within will be automatically run by [Tests/RunTests.jsl#L43](https://github.com/sassoftware/jsl-hamcrest/blob/master/Tests/RunTests.jsl#L43)

In addition, this work needs to go under a new `1.2.0` `Added` section in the changelog. I forgot to do that when we released `1.1.0`. I am adding this in #61

emccorkle commented on Source/Reporters/JunitXMLReporter.jsl:
```diff
@@ -0,0 +1,247 @@
+				Char( this:successes[i][2] )
+			) || " ";
+			tmp_label = this:successes[i][1];
+			list_tmp_label = words(tmp_label, "⮚");
```
This should be [ut concat test label sep](https://github.com/sassoftware/jsl-hamcrest/blob/5a0d45b6341a947056edd250c3484be18cf77b5f/Source/Core/Utils.jsl#L466-L467) everywhere.

dougiechong replied:
okay, I will make that change

dougiechong committed:
replace hard coded delimiter with variable

emccorkle commented on Source/Reporters/JunitXMLReporter.jsl:
```diff
@@ -0,0 +1,247 @@
+				Char( this:failures[i][2] )
+			) || " ";
+			tmp_label = this:failures[i][1];
+			list_tmp_label = words(tmp_label, "⮚");
```
Unfortunately, if the `ut assert that` is run outside of a `ut test`, there may not be three elements in this `list_tmp_label`. See [UtDetailsTreeReporter](https://github.com/sassoftware/jsl-hamcrest/blob/289124e570371580fe2ad95f879284e7d0740139/Source/Addin/Reporters.jsl#L280-L289) and #59.

dougiechong replied:
Should be handled in commit b0d89cefc383a60df3a7732de648366d8bd621cd. I still need to handle labels that could be 4 elements, but now 1 and 3 should work fine. The main idea of the solution is to put assertions outside of test cases in an anonymous test suite and numbered anonymous test cases if no label is supplied at all.

dougiechong replied:
4 part labels are also handled now

emccorkle commented on Source/Reporters/JunitXMLReporter.jsl:
```diff
@@ -0,0 +1,247 @@
+		If(content != "",
+			// replace all newline with doublespace then newline
+			Substitute Into(content, "\!n", "\!n  ");
+			xml_tag ||= "  " || content || "\!n";
```
`vals` and `content` often both come from human-readable free-form text. It may be part of an expected or actual value. It may be written by a matcher. It may be written by a user as a test case name, test name, or assertion label. Or something else. Please use `XML Encode` for attribute values and XML text to ensure they don't contain invalid characters or prematurely end the attribute or node.

emccorkle replied:
I added some failing tests in https://github.com/sassoftware/jsl-hamcrest/pull/60/commits/6346b9fe8356a2e990bf91c6c2a7068106aa0dea for this.

dougiechong replied:
fixed to escape non valid xml characters in commit d282dd545646455c25c93815c68331cc840b0e6d

JustinChilton commented on Source/Reporters/JunitXMLReporter.jsl:
```diff
@@ -0,0 +1,247 @@
+				
+			);
+		);
+		//Write( "\!n--------- Unexpected Throws ----------" );
```
Remove commented code.

JustinChilton commented on Source/Reporters/JunitXMLReporter.jsl:
```diff
@@ -0,0 +1,247 @@
+			</testsuite>
+		</testsuites>
+        ---------
+*/
```
Since this is part of a class, you can indent the documentation over to be even with the method definition. 

Also, you can use `Method: show_xml_report` instead of `Function: show_xml_report`. Both are keywords in natural docs, but `Method` is more specific.

JustinChilton replied:
Same thing with `generate_xml_tag`

dougiechong replied:
Do you mean just indenting the opening and closing comments? The rest is already indented with the method. I'll update Function to Method

JustinChilton replied:
I think it would be best to indent the whole comment block over. It is a little confusing since it looks like there is a mix of tabs and spaces in there. It should look something like [this](https://github.com/sassoftware/jsl-hamcrest/blob/master/Source/Core/Core.jsl#L119-L141) where Function/Arguments/Example line up, but just tabbed over once to line up with the method definition.

dougiechong replied:
should be fixed in commit 212ea1a. Let me know if 4 spaces for indents is required and I will change accordingly

emccorkle reviewed (commented):
I'm interested in how `unexpected throw` is being reported. Does JUnit or XUnit make a distinction between these and assertion failures? In addition, have you tried any downstream consumers of this xml yet? Jenkins or some other JUnit output renderer?

JustinChilton commented on Source/Reporters/JunitXMLReporter.jsl:
```diff
@@ -0,0 +1,247 @@
+	//
+	// Dump a report to the log detailing all the successes and
+	// failures for each test in JunitXML format shown in the comments at the top.
+	show report = Method( {},
```
Since this `show report` method has a lot of logic to create the result string, could we change this to `get report` that returns the string and add `show report` that just calls `Write(this:get report())`? This would allow you to do other things with the report, like write it to a file.

JustinChilton replied:
This was resolved by @emccorkle's commit.

JustinChilton commented on Source/Reporters/JunitXMLReporter.jsl:
```diff
@@ -0,0 +1,247 @@
+		</testsuites>
+        ---------
+*/
+	show_xml_report = Method({xml_results},
```
Maybe change the name of this method to `get_xml_report`? It does not do any showing, just returns a string.

dougiechong replied:
sounds good, I will make the change

dougiechong replied:
Done in commit 846e93f3706107d35a48fd74d83ad7285baa46c2

emccorkle reviewed (commented):
Could you please add a factory function to this file (see [ut collecting reporter](https://github.com/sassoftware/jsl-hamcrest/blob/master/Source/Reporters/CollectingReporter.jsl#L121-L125))?

emccorkle committed:
Added some tests for `UtJunitXMLReporter`.

Also added `ut valid xml` supporting matcher.

Signed-off-by: Evan McCorkle <Evan.McCorkle@jmp.com>

emccorkle commented:
Since there was no `Tests/UnitTests/Reporters` folder and no substantial examples of `Reporter` tests, I added some to this PR. Hopefully this doesn't cause you any trouble. See https://github.com/sassoftware/jsl-hamcrest/pull/60/commits/6346b9fe8356a2e990bf91c6c2a7068106aa0dea for details. In addition, this adds a new `ut valid xml` matcher.

dougiechong commented:
I appreciate all the feedback and the example tests, but I am unable to run them successfully. I get the following error when I try to run the test:

Unexpected "n root tests". Perhaps there is a missing ";" or ",".
Line 4 Column 1: ►n root tests = Function({xml}, {out = .},

It seems like the comments at the top cause this problem and removing them allows the code to run

dougiechong committed:
change function comments to method
remove unnessesary comment

dougiechong committed:
replace all tabs with doublespace

dougiechong committed:
add factory function for junitxml reporter

dougiechong commented:
> Could you please add a factory function to this file (see [ut collecting reporter](https://github.com/sassoftware/jsl-hamcrest/blob/master/Source/Reporters/CollectingReporter.jsl#L121-L125))?

Done in commit 
c29793e

dougiechong committed:
add a simple test for reporting a success
also added a test for one failure

dougiechong committed:
remove all tabs, comments should line up now

dougiechong committed:
rename show_xml_report to get_xml_report

dougiechong committed:
add parse_label method
add the method parse label to handle labels not generated by ut test.
Currently will only support labels with 3 sections for test suite,
test case and assertion number, or with one section as an anonymous test
This change will allow reporting of assertions not created by ut test

dougiechong committed:
remove accidental show statement

dougiechong committed:
split setup expr into it's own expression
add a test for single failing assertion

dougiechong committed:
add escape content flag
Add the flag to optional escape &,<,and > in the content of the xml tag.
Use XML encode to always escape
value attributes and names

dougiechong committed:
pull num outside of parse XML
when num is called inside parse xml it simply counts number of
attirbutes

dougiechong committed:
put num back in parse xml
simply needed to put setup_expr in an expr() statement, tests were being
double counted

dougiechong committed:
number tests in comments
convert simple tests to wtih junit reporter style

emccorkle committed:
Adjusted doc syntax to work with NaturalDocs.

Signed-off-by: Evan McCorkle <Evan.McCorkle@jmp.com>

emccorkle committed:
Fixed test failure when running from RunTests.jsl.

Multiline string literal has different line endings in the editor vs
when being included. Tests would pass when run directly from
JunitXMLReporterTest but fail when run from RunTests. Make
sure both line endings get mapped to .+

Signed-off-by: Evan McCorkle <Evan.McCorkle@jmp.com>

dougiechong committed:
add ability to handle labels with 4 parts

dougiechong committed:
add 4 part label test and a multiple success/fail

emccorkle committed:
Added tests for unexpected throws and suite names.

Signed-off-by: Evan McCorkle <Evan.McCorkle@jmp.com>

dougiechong committed:
merge accepting \!N as well as \!r

emccorkle reviewed (changes requested):
Could you please add the two line Copyright and License header to the JunitXMLReporter?

dougiechong committed:
merge test for throw

emccorkle commented on Source/Reporters/JunitXMLReporter.jsl:
```diff
@@ -0,0 +1,321 @@
+  //  hash_results - Associative Array with test suites as keys and associative arrays with test cases as keys as values
+  // 
+  // Results:
+  //  The label split by the separator. The second and third values will be joined by an _
```
Any reason not to join with `ut concat test label sep` since we already have that restriction?

dougiechong replied:
it could not recognize the variable "ut concat test label sep" unless I passed it to the method as an argument. I'm not sure why

emccorkle replied:
`Methods` are implicitly `Default Local`. So it probably thought that was a local. Try explicitly scoping it with `::ut concat test label sep`.

dougiechong replied:
That does seem to work, however I am curious as to why it is not required in the get report method

dougiechong committed:
add two line copyright and a comment back

emccorkle commented on Tests/UnitTests/Reporters/JunitXMLReporterTest.jsl:
```diff
@@ -0,0 +1,421 @@
+	";
+
+	//replace tabs and carriage returns used for easier readability
+	Substitute Into(pattern, "\!r", ".+", "\!t", "");
```
This test is failing when run from `RunTests.jsl`. It needs the \!N replacement.

dougiechong replied:
fixed in commit 7dda28f6a337f12e2526a4b768e7eab2f21e3201

dougiechong commented:
> Could you please add the two line Copyright and License header to the JunitXMLReporter?

done in commit 7ccce9af59f587d1a1d2af9887fcd3843b2f8a28

dougiechong committed:
add "\!N" to make run tests pass

dougiechong commented:
@emccorkle @JustinChilton, I believe I have made all the requested changes, including formatting, escaping invalid xml characters, and handling labels of length 1,3, and 4.

Please let me know if any other changes are required to merge this pull request.

Handling throws separately can be marked as an improvement for a later time

dougiechong committed:
remove the need for sep as an argument
argument was no longer needed as the global
separtor ::ut concat test label sep could be used

dougiechong committed:
remove the need for sep as an argument
argument was no longer needed as the global
separtor ::ut concat test label sep could be used

dougiechong committed:
Merge branch 'add-Junit-xml-reporter' of https://github.com/dougiechong/jsl-hamcrest into add-Junit-xml-reporter
